<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>A Flexible Generalized Linear Mixed Effects Model for Testing Cell-Cell Colocalization in Spatial Immunofluorescent Data</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Statistical analysis of spatial single-cell protein data</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Workshop
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="fridley.html">Overview of multiplex imaging analysis techniques</a>
    </li>
    <li>
      <a href="vandekar.html">Normalization and cell phenotyping for mIF data</a>
    </li>
    <li>
      <a href="chervoneva.html">Quantile index biomarkers for mIF data</a>
    </li>
    <li>
      <a href="wrobel.html">Tools and software for FDA of mIF data</a>
    </li>
    <li>
      <a href="ma.html">A flexible GLM for testing cell-cell colocalization in mIF data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="https://github.com/julia-wrobel/PSB_scProteomics">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">A Flexible Generalized Linear Mixed Effects
Model for Testing Cell-Cell Colocalization in Spatial Immunofluorescent
Data</h1>

</div>


<div id="slide-deck" class="section level2">
<h2>Slide Deck</h2>
<iframe class="speakerdeck-iframe" frameborder="0" src="https://speakerdeck.com/player/90ae3791ec61459f8cf52e502486322c" title="psb_ma" allowfullscreen="true" style="border: 0px; background: padding-box padding-box rgba(0, 0, 0, 0.1); margin: 0px; padding: 0px; border-radius: 6px; box-shadow: rgba(0, 0, 0, 0.2) 0px 5px 40px; width: 100%; height: auto; aspect-ratio: 560 / 315;" data-ratio="1.7777777777777777">
</iframe>
</div>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>This module focuses on testing cell-cell interaction (colocalization)
using spatial regression, implemented in the R package
<code>spaMM</code>.</p>
<p>Load data and libraries.</p>
<pre class="r"><code># required packages
library(tidyverse)
library(spaMM)
theme_set(theme_bw() + theme(legend.position = &quot;bottom&quot;))

# load processed ovarian cancer data
load(url(&quot;https://github.com/julia-wrobel/MI_tutorial/raw/main/Data/ovarian.RDA&quot;))</code></pre>
</div>
<div id="visualize-cell-colocalization" class="section level2">
<h2>Visualize cell colocalization</h2>
<p>We visualize two samples from the ovarian cancer dataset from <a
href="https://bioconductor.org/packages/release/data/experiment/html/VectraPolarisData.html">VectraPolarisData</a>
and colocalization of B-cells and CD8- T-cells. We are interested in
testing if this colocalization is significant (i.e., more interaction
than random chance), and if it correlates with five-year survival.</p>
<pre class="r"><code># create five-year survival variable
ovarian_df &lt;- ovarian_df %&gt;%
  mutate(vysurvival = case_when(
    survival_time &gt;= 60  ~ 1,
    death == 1 ~ 0,
    TRUE ~ NA_real_
  ))

# cell phenotype variable
ovarian_df &lt;- ovarian_df %&gt;%
  mutate(phenotype = case_when(
    phenotype_cd68 == &quot;CD68+&quot; ~ &quot;macrophage&quot;,
    phenotype_cd19 == &quot;CD19+&quot; ~ &quot;B-cell&quot;,
    phenotype_cd3 == &quot;CD3+&quot; &amp; phenotype_cd8 == &quot;CD8+&quot; ~ &quot;CD8+ T-cell&quot;,
    phenotype_cd3 == &quot;CD3+&quot; &amp; phenotype_cd8 == &quot;CD8-&quot; ~ &quot;CD8- T-cell&quot;,
    TRUE ~ &quot;other&quot;
  )) 

# subset to two samples with moderate B-cell populations
subjs_subset &lt;- ovarian_df %&gt;% 
  filter(!is.na(vysurvival)) %&gt;% 
  group_by(sample_id, vysurvival) %&gt;% 
  filter(any(phenotype == &quot;macrophage&quot;), 
         any(phenotype == &quot;B-cell&quot;),
         any(phenotype == &quot;CD8+ T-cell&quot;),
         any(phenotype == &quot;CD8- T-cell&quot;)) %&gt;%
  summarise(n_b_cells = sum(phenotype == &quot;B-cell&quot;)) %&gt;% 
  filter(n_b_cells &lt; 100) %&gt;% 
  group_by(vysurvival) %&gt;% 
  arrange(-n_b_cells) %&gt;%
  slice(1) %&gt;%
  pull(sample_id)

ovarian_df_subset &lt;- ovarian_df %&gt;% 
  filter(sample_id %in% subjs_subset)

# visualize B-cells and CD8- T-cells across images
p_viz &lt;- ovarian_df_subset %&gt;% 
  group_by(sample_id) %&gt;% 
  mutate(x_viz = x - min(x),
                y_viz = y - min(y)) %&gt;% 
  mutate(phenotype_viz = phenotype %&gt;% 
                  recode(
                    &quot;CD8+ T-cell&quot; = &quot;other&quot;,
                    &quot;macrophage&quot; = &quot;other&quot;) %&gt;% 
                  factor(levels = c(&quot;other&quot;, &quot;CD8- T-cell&quot;, &quot;B-cell&quot;))) %&gt;%
  arrange(phenotype_viz) %&gt;% 
  ggplot(aes(x_viz, y_viz, color = phenotype_viz, size = phenotype_viz)) +
  geom_point(alpha = 0.5) +
  scale_color_manual(values = c(&quot;B-cell&quot; = &quot;red&quot;, 
                                &quot;CD8- T-cell&quot; = &quot;blue&quot;, 
                                &quot;other&quot; = &quot;grey&quot;)) +
  scale_size_manual(values = c(&quot;B-cell&quot; = 0.5, 
                                &quot;CD8- T-cell&quot; = 0.5, 
                                &quot;other&quot; = 0.2)) +
  facet_wrap(~ paste0(&quot;vysurvival &quot;, vysurvival) + 
               paste0(&quot;sample_id &quot;, sample_id), nrow = 1) +
  # annotation for 25 micron scale bar
  annotate(geom = &quot;segment&quot;, x = 10, y = 10, xend = 35, yend = 10) +
  annotate(geom = &quot;text&quot;, x = 22.5, y = 10, label = &quot;25&quot;, 
           hjust = 0.5, vjust = -0.5) +
  coord_fixed() +
  theme(legend.position =&quot;bottom&quot;,
        axis.text = element_blank(),
        axis.ticks = element_blank(), 
        panel.grid = element_blank())
print(p_viz)</code></pre>
<p><img src="ma_files/figure-html/unnamed-chunk-2-1.png" width="90%" /></p>
</div>
<div id="permutation-test-for-cell-colocalization"
class="section level2">
<h2>Permutation test for cell colocalization</h2>
<p>Here we adopt a permutation based approach to test for cell-cell
interaction. This was used in e.g. <a
href="https://pubmed.ncbi.nlm.nih.gov/28783155/">histoCAT</a>.</p>
<ul>
<li><p>Red vertical lines indicate observed number of B-cell/CD8- T-cell
interactions in each image.</p></li>
<li><p>Histogram indicates number of interactions from randomly
permuting cell phenotype labels.</p></li>
<li><p>This approach can test for enrichment of interactions in each
image, but is difficult to generalize across images, and to test for
interaction correlation with phenotypes/outcomes.</p></li>
</ul>
<pre class="r"><code>threshold &lt;- 25

# construct adjacency matrices for each image
l_mat_adj &lt;- ovarian_df_subset %&gt;% 
  group_split(sample_id) %&gt;% 
  map(function(i_ovarian_df) {
    mat_dist &lt;- i_ovarian_df %&gt;% 
      select(x, y) %&gt;% 
      as.matrix() %&gt;% 
      dist() %&gt;% 
      as.matrix()
    dimnames(mat_dist) &lt;- list(i_ovarian_df$cell_id, i_ovarian_df$cell_id)
    mat_adj &lt;- mat_dist &lt; threshold 
    # don&#39;t count cell as its own neighbor
    diag(mat_adj) &lt;- 0
    
    return(mat_adj)
  })

# visualize permuted interaction vs observed interaction
n_perm &lt;- 500
set.seed(0)
df_interactions &lt;- ovarian_df_subset %&gt;% 
  group_split(sample_id) %&gt;% 
  map2_dfr(l_mat_adj, function(i_ovarian_df, i_mat_adj) {
    n_observe &lt;- sum(i_mat_adj[i_ovarian_df$phenotype == &quot;B-cell&quot;, 
                                i_ovarian_df$phenotype == &quot;CD8- T-cell&quot;])
    n_permute &lt;- seq(1, n_perm) %&gt;% 
      map_dbl(function(i_perm) {
        i_ovarian_df_perm &lt;- i_ovarian_df %&gt;% 
          sample_n(size = nrow(.), replace = FALSE)
        sum(i_mat_adj[i_ovarian_df_perm$phenotype == &quot;B-cell&quot;, 
                       i_ovarian_df_perm$phenotype == &quot;CD8- T-cell&quot;])
      })
    tibble(
      sample_id = i_ovarian_df$sample_id[1],
      vysurvival = i_ovarian_df$vysurvival[1],
      n_observe = n_observe,
      n_permute = n_permute
    )
  })

p_interactions &lt;- df_interactions %&gt;% 
  ggplot(aes(x = n_permute)) +
  geom_histogram(bins = 50) +
  geom_text(data = df_interactions %&gt;% filter(!duplicated(sample_id)),
             aes(label = paste0(&quot;#interactions from data=&quot;, n_observe)), 
            x = Inf, y = Inf,
            vjust = 1.2, hjust = 1.05,
            color = &quot;red&quot;) +
  geom_vline(data = df_interactions %&gt;% 
               filter(sample_id == 26) %&gt;% 
               filter(!duplicated(sample_id)),
             aes(xintercept = n_observe), color = &quot;red&quot;) +
  facet_wrap(~ paste0(&quot;vysurvival &quot;, vysurvival) + 
               paste0(&quot;sample_id &quot;, sample_id), nrow = 1) +
  xlab(&quot;#interactions from permutation&quot;)
print(p_interactions)</code></pre>
<p><img src="ma_files/figure-html/unnamed-chunk-3-1.png" width="90%" /></p>
</div>
<div id="test-for-cell-colocalization-using-spatial-regression"
class="section level2">
<h2>Test for cell colocalization using spatial regression</h2>
<p>We can perform spatial regression on each B-cell’s local
neighborhood, and the enrichment of CD8 T-cells in each
neighborhood.</p>
<div id="visualize-the-local-neighborhood-of-one-b-cell"
class="section level3">
<h3>Visualize the local neighborhood of one B-cell</h3>
<pre class="r"><code># calculate number of neighboring cells for each B-cell
df_nb &lt;- ovarian_df_subset %&gt;% 
  group_split(sample_id) %&gt;% 
  map2_dfr(l_mat_adj, function(i_ovarian_df, i_mat_adj) {
    # count the number of neighbors for each B-cell
    i_df_nb &lt;- i_mat_adj[i_ovarian_df$phenotype == &quot;B-cell&quot;, ] %&gt;% 
      apply(1, function(x) tapply(x, i_ovarian_df$phenotype, sum)) %&gt;% 
      t() %&gt;% 
      as.data.frame() %&gt;% 
      rownames_to_column(&quot;cell_id&quot;) %&gt;% 
      mutate(total = `B-cell` + `CD8- T-cell` + `CD8+ T-cell` + macrophage + other,
             sample_id = i_ovarian_df$sample_id[1]) %&gt;% 
      # remove singletons
      filter(total &gt; 0)
    
    return(i_df_nb)
  })
df_nb &lt;- df_nb %&gt;% 
  mutate(cell_id = as.numeric(cell_id)) %&gt;% 
  left_join(ovarian_df_subset, 
            by = c(&quot;sample_id&quot;, &quot;cell_id&quot;))

# visualize the local neighborhood of a particular B-cell
df_one_cell &lt;- df_nb %&gt;% 
  filter(sample_id == 25) %&gt;% 
  arrange(-`CD8- T-cell`) %&gt;% 
  slice(1)
nbs &lt;- l_mat_adj[[1]][as.character(df_one_cell$cell_id), ] %&gt;% 
  keep(~.x &gt; 0) %&gt;% 
  names() %&gt;% 
  as.numeric()
nbs &lt;- c(nbs, df_one_cell$cell_id)
df_cell_viz &lt;- ovarian_df_subset %&gt;% 
  filter(sample_id == 25) %&gt;% 
  filter(x &lt; df_one_cell$x + 30 &amp; x &gt; df_one_cell$x - 30,
         y &lt; df_one_cell$y + 30 &amp; y &gt; df_one_cell$y - 30)
p_cell_viz &lt;- df_cell_viz %&gt;%
  mutate(phenotype_viz = phenotype %&gt;% 
                  recode(
                    &quot;CD8+ T-cell&quot; = &quot;other&quot;,
                    &quot;macrophage&quot; = &quot;other&quot;) %&gt;% 
                  factor(levels = c(&quot;other&quot;, &quot;CD8- T-cell&quot;, &quot;B-cell&quot;))) %&gt;% 
  ggplot(aes(x = x, y = y, 
             color = phenotype_viz, 
             size = cell_id == df_one_cell$cell_id)) +
  geom_point() +
  scale_color_manual(values = c(&quot;B-cell&quot; = &quot;red&quot;, &quot;CD8- T-cell&quot; = &quot;blue&quot;,
                                &quot;other&quot; = &quot;grey&quot;)) +
  # don&#39;t show size legend
  scale_size_manual(values = c(&quot;TRUE&quot; = 3, &quot;FALSE&quot; = 2), guide = &quot;none&quot;) +
  # draw a neighborhood circle around the center cell
  annotate(&quot;path&quot;,
           x = df_one_cell$x + 25*cos(seq(0,2*pi,length.out=100)),
           y = df_one_cell$y + 25*sin(seq(0,2*pi,length.out=100)),
           color = &quot;red&quot;, 
           linetype = &quot;dashed&quot;) +
  coord_fixed() +
  theme(axis.ticks = element_blank(),
        axis.text = element_blank(),
        panel.grid = element_blank())
print(p_cell_viz)</code></pre>
<p><img src="ma_files/figure-html/unnamed-chunk-4-1.png" width="90%" /></p>
<pre class="r"><code>
# calculate overall density of cells within each image
# used as &quot;offset&quot; terms in spaMM regression
df_density &lt;- ovarian_df_subset %&gt;%
  group_by(sample_id) %&gt;% 
  mutate(n_total = n()) %&gt;% 
  group_by(sample_id, phenotype) %&gt;% 
  summarise(density = n() / n_total[1]) %&gt;% 
  pivot_wider(names_from = phenotype, values_from = density, 
                     names_prefix = &quot;density_&quot;)
df_nb &lt;- df_nb %&gt;% 
  left_join(df_density, by = &quot;sample_id&quot;)</code></pre>
</div>
<div id="spamm-regression" class="section level3">
<h3><code>spaMM</code> regression</h3>
<p>Fit spatial regression using <code>spaMM::fitme</code>. Regression
terms:</p>
<ul>
<li><p><code>vysurvival</code>: the variable of interest.</p></li>
<li><p><code>Matern(1|x + y %in% sample_id)</code>: Matern spatial
correlation term.</p></li>
<li><p><code>offset(log(\`density_CD8- T-cell` / (1 - `density_CD8- T-cell`)))</code>:
offset term for “background” prevalence of CD8- T-cells.</p></li>
</ul>
<pre class="r"><code>fit_spaMM &lt;- 
  fitme(
    formula = 
      cbind(`CD8- T-cell`, total - `CD8- T-cell`) ~ 
      # covariate of interest
      vysurvival + 
      # spatial correlation term
      Matern(1|x + y %in% sample_id) +
      # offset term for &quot;background&quot; prevalence of CD8- T-cells
      offset(log(`density_CD8- T-cell` / (1 - `density_CD8- T-cell`))),
    data = df_nb,
    family = binomial(link = &quot;logit&quot;),
    fixed = list(&quot;nu&quot; = 0.5),
    method = &quot;ML&quot;)
print(fit_spaMM)
## formula: cbind(`CD8- T-cell`, total - `CD8- T-cell`) ~ vysurvival + Matern(1 | 
##     x + y %in% sample_id) + offset(log(`density_CD8- T-cell`/(1 - 
##     `density_CD8- T-cell`)))
## Estimation of corrPars and lambda by ML (p_v approximation of logL).
## Estimation of fixed effects by ML (p_v approximation of logL).
## Estimation of lambda by &#39;outer&#39; ML, maximizing logL.
## family: binomial( link = logit ) 
##  ------------ Fixed effects (beta) ------------
##             Estimate Cond. SE t-value
## (Intercept)    2.303   0.3996   5.764
## vysurvival    -2.002   0.6039  -3.316
##  --------------- Random effects ---------------
## Family: gaussian( link = identity ) 
##                    --- Correlation parameters:
##       1.nu      1.rho 
## 0.50000000 0.01121604 
##            --- Variance parameters (&#39;lambda&#39;):
## lambda = var(u) for u ~ Gaussian; 
##    x + y %in.  :  1.116  
## # of obs: 171; # of groups: x + y %in., 171 
##  ------------- Likelihood values  -------------
##                         logLik
## logL       (p_v(h)): -161.5562</code></pre>
</div>
<div id="understanding-the-output-of-fitme" class="section level3">
<h3>Understanding the output of <code>fitme</code></h3>
<p>Many of these are detailed in the documentation
<code>?HLfit</code>.</p>
<ul>
<li><p>Fixed effects: estimated coefficients. The conditional standard
errors are not very informative, as they do not account for the
uncertainty in the random effects term. Below we will perform
inferenence on these parameters using likelihood ratio tests
instead.</p></li>
<li><p>Random effects:</p>
<ul>
<li><p>Matern correlation parameters (see <code>?MaternCorr</code> for
more details):</p>
<ul>
<li><p><code>nu</code>: smoothness parameter for the Matern correlation
function. We pre-fixed this at 1/2.</p></li>
<li><p><code>rho</code>: estimated correlation length for the Matern
correlation function. This is the parameter that controls the range of
the spatial correlation.</p></li>
<li><p>Under these specifications, the Matern correlation for two cells
with distance <code>d</code> is <code>exp(-rho * d)</code>. See
<code>?MaternCorr</code> for more details.</p></li>
</ul></li>
<li><p>Variance of spatial correlation (<code>lambda</code>):</p>
<ul>
<li><p>Details in <code>?`random-effects`</code>.</p></li>
<li><p><code>lambda</code> is the variance of i.i.d normal random
variables, which are then multipled by a Cholesky decomposition of the
Matern correlation to generate spatially correlated random
effects.</p></li>
</ul></li>
</ul></li>
</ul>
</div>
<div id="lrt-testing-of-fixed-effects" class="section level3">
<h3>LRT testing of fixed effects</h3>
<p>We perform LRT on fixed effects by contrasting the fully specified
model versus “null” models.</p>
<ul>
<li>The intercept is the enrichment of CD8- T-cell colocalizaiton in
B-cell neighborhoods (when <code>vysurvival = 0</code>), accounting for
the “background” chance of CD8- T-cells in the images.</li>
</ul>
<pre class="r"><code>fit_spaMM_null1 &lt;- 
  fitme(
    formula = 
      cbind(`CD8- T-cell`, total - `CD8- T-cell`) ~ 
      # null model with no intercept
      0 + vysurvival +
      Matern(1|x + y %in% sample_id) +
      offset(log(`density_CD8- T-cell` / (1 - `density_CD8- T-cell`))),
    data = df_nb,
    family = binomial(link = &quot;logit&quot;),
    fixed = list(&quot;nu&quot; = 0.5),
    method = &quot;ML&quot;)

1 - pchisq(2 * (logLik(fit_spaMM) - logLik(fit_spaMM_null1)), df = 1)
##         p_v 
## 0.002247611</code></pre>
<ul>
<li>The coefficient for <code>vysurvival</code> is whether or not this
enrichment is correlated with five-year survival.</li>
</ul>
<pre class="r"><code>fit_spaMM_null2 &lt;- 
  fitme(
    formula = 
      cbind(`CD8- T-cell`, total - `CD8- T-cell`) ~ 
      # null model with no vysurvival
      1 + 
      Matern(1|x + y %in% sample_id) +
      offset(log(`density_CD8- T-cell` / (1 - `density_CD8- T-cell`))),
    data = df_nb,
    family = binomial(link = &quot;logit&quot;),
    fixed = list(&quot;nu&quot; = 0.5),
    method = &quot;ML&quot;)

1 - pchisq(2 * (logLik(fit_spaMM) - logLik(fit_spaMM_null2)), df = 1)
##         p_v 
## 0.007631282</code></pre>
</div>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<ul>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/34529988/">Colocalization
of Treg with other immune cells in the IBD epithelium</a></li>
<li><a
href="https://pubmed.ncbi.nlm.nih.gov/28783155/">histoCAT</a></li>
<li><a
href="https://cran.r-project.org/web/packages/spaMM/index.html">spaMM</a></li>
<li><a
href="https://bioconductor.org/packages/release/data/experiment/html/VectraPolarisData.html">VectraPolarisData</a></li>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/34615692/">Ovarian cancer
paper</a></li>
</ul>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
