<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>VectraPolarisData: experiment-scale multiplex imaging data</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.2.1/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.2.1/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Analysis of multiplex single-cell imaging data in R</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Short Course
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="MI_Data.html">Data Download</a>
    </li>
    <li>
      <a href="MI_intro.html">Background on multiplex single-cell imaging</a>
    </li>
    <li>
      <a href="MI_VectraPolarisData.html">VectraPolarisData package</a>
    </li>
    <li>
      <a href="MI_normalization.html">Segmentation, normalization, and phenotyping</a>
    </li>
    <li>
      <a href="MI_spatialAnalysis.html">Extracting spatial summary statistics</a>
    </li>
    <li>
      <a href="MI_fda.html">Modeling patient-level outcomes</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="https://github.com/julia-wrobel/MI_tutorial">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">VectraPolarisData: experiment-scale
multiplex imaging data</h1>

</div>


<div id="slide-deck" class="section level2">
<h2>Slide Deck</h2>
<iframe class="speakerdeck-iframe" frameborder="0" src="https://speakerdeck.com/player/5573ed99b24a464885bbbab47af138d5" title="MI_VectraPolarisData" allowfullscreen="true" style="border: 0px; background: padding-box padding-box rgba(0, 0, 0, 0.1); margin: 0px; padding: 0px; border-radius: 6px; box-shadow: rgba(0, 0, 0, 0.2) 0px 5px 40px; width: 100%; height: auto; aspect-ratio: 560 / 314;" data-ratio="1.78343949044586">
</iframe>
</div>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>The <a
href="https://bioconductor.org/packages/release/data/experiment/html/VectraPolarisData.html">VectraPolarisData</a>
<a
href="http://bioconductor.org/packages/release/bioc/html/ExperimentHub.html">ExperimentHub</a>
package provides two large multiplex immunofluorescence datasets
collected using Akoya Biosciences Vectra 3 and Vectra Polaris platforms.
Image preprocessing (cell segmentation and phenotyping) was performed
using <a
href="https://www.akoyabio.com/phenoimager/software/inform-tissue-finder/">inForm</a>
software. Data are provided as objects of class <a
href="https://bioconductor.org/packages/release/bioc/html/SpatialExperiment.html">SpatialExperiment</a>.</p>
<pre class="r"><code># load libraries
library(tidyverse)</code></pre>
</div>
<div id="install-and-load-data" class="section level2">
<h2>Install and load data</h2>
<p>The data is publicly available on Bioconductor as the package
<code>VectraPolarisData</code>. You can install the package from
Bioconductor here:</p>
<pre class="r"><code>if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE)) {
     install.packages(&quot;BiocManager&quot;)
}

BiocManager::install(&quot;VectraPolarisData&quot;)
</code></pre>
</div>
<div id="ovarian-cancer-data" class="section level2">
<h2>Ovarian cancer data</h2>
<p>Load the high grade serous ovarian cancer data. This dataset has been
segmented and phenotyped, and has one image per patient.</p>
<pre class="r"><code>library(VectraPolarisData)

# load spatial experiment object
oc &lt;- HumanOvarianCancerVP()</code></pre>
<p>Inspect the <code>SpatialExperiment</code> object.</p>
<pre class="r"><code># check object
oc
## class: SpatialExperiment 
## dim: 10 1610431 
## metadata(1): clinical_data
## assays(3): intensities nucleus_intensities membrane_intensities
## rownames(10): ck_opal_780 ki67_opal_690 ... dapi autofluorescence
## rowData names(0):
## colnames: NULL
## colData names(195): cell_id tissue_category ... phenotype_cd8 sample_id
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## spatialCoords names(2) : cell_x_position cell_y_position
## imgData names(0):</code></pre>
<p>The object stores the following marker values for each marker and
each cell</p>
<ul>
<li><code>intensities</code>: mean marker intensity for entire cell</li>
<li><code>nucleus_intensities</code>: mean marker intensity in the
nucleus</li>
<li><code>membrane_intensities</code>: mean marker intensity in the cell
membrane</li>
</ul>
<pre class="r"><code>assayNames(oc)
## [1] &quot;intensities&quot;          &quot;nucleus_intensities&quot;  &quot;membrane_intensities&quot;</code></pre>
<p>Row names provide the names of the markers.</p>
<pre class="r"><code>rownames(oc)
##  [1] &quot;ck_opal_780&quot;      &quot;ki67_opal_690&quot;    &quot;cd8_opal_650&quot;     &quot;ier3_opal_620&quot;   
##  [5] &quot;p_stat3_opal_570&quot; &quot;cd3_opal_540&quot;     &quot;cd68_opal_520&quot;    &quot;cd19_opal_480&quot;   
##  [9] &quot;dapi&quot;             &quot;autofluorescence&quot;</code></pre>
<p>The colData slot contains cell-level covariates like cell phenotype,
cell area, and, min, max, and standard deviation of intensity values for
each marker.</p>
<pre class="r"><code>dim(colData(oc))
## [1] 1610431     195</code></pre>
<p>Many of the variables are summary statistics of marker intensity
values. More information on the different variables in this and the lung
cancer dataset can be found in the <a
href="https://bioconductor.org/packages/release/data/experiment/vignettes/VectraPolarisData/inst/doc/VectraPolarisData.html">package
vignette</a>.</p>
<pre class="r"><code>set.seed(12333)
sample(names(colData(oc)), 15)
##  [1] &quot;entire_cell_cd3_opal_540_std_dev&quot;   &quot;nucleus_ki67_opal_690_std_dev&quot;     
##  [3] &quot;entire_cell_dapi_dapi_min&quot;          &quot;entire_cell_cd3_opal_540_min&quot;      
##  [5] &quot;cytoplasm_cd68_opal_520_total&quot;      &quot;cytoplasm_dapi_dapi_min&quot;           
##  [7] &quot;entire_cell_cd8_opal_650_max&quot;       &quot;entire_cell_dapi_dapi_total&quot;       
##  [9] &quot;cytoplasm_p_stat3_opal_570_std_dev&quot; &quot;membrane_ier3_opal_620_std_dev&quot;    
## [11] &quot;entire_cell_cd3_opal_540_total&quot;     &quot;entire_cell_p_stat3_opal_570_max&quot;  
## [13] &quot;nucleus_cd3_opal_540_total&quot;         &quot;membrane_p_stat3_opal_570_max&quot;     
## [15] &quot;nucleus_cd68_opal_520_min&quot;</code></pre>
<div id="converting-to-dataframe" class="section level3">
<h3>Converting to dataframe</h3>
<p>Storing the data as a <code>SpatialExperiment</code> allows for
analysis using tools that are built into the
<code>SpatialExperiment</code> workflow. However, it is often useful to
work with the data as a <code>data.frame</code> object instead.</p>
<p>Code below converts the ovarian cancer dataset from a
<code>SpatialExperiment</code> object to a <code>data.frame</code>. We
perform some data cleaning steps as well.</p>
<pre class="r"><code>
## Assays slots
assays_slot &lt;- assays(oc)
intensities_df &lt;- assays_slot$intensities
nucleus_intensities_df&lt;- assays_slot$nucleus_intensities
rownames(nucleus_intensities_df) &lt;- paste0(&quot;nucleus_&quot;, rownames(nucleus_intensities_df))
membrane_intensities_df&lt;- assays_slot$membrane_intensities
rownames(membrane_intensities_df) &lt;- paste0(&quot;membrane_&quot;, rownames(membrane_intensities_df))

# colData and spatialData
colData_df &lt;- colData(oc)
spatialCoords_df &lt;- spatialCoords(oc)

# clinical data
patient_level_ovarian &lt;- metadata(oc)$clinical_data %&gt;%
  # create binary stage variable
  dplyr::mutate(stage_bin = ifelse(stage %in% c(&quot;1&quot;, &quot;2&quot;), 0, 1))

# only include samples for whom we have an id in the patient dataset, which is 128- the other 4 are controls
# give Markers shorter names
cell_level_ovarian &lt;- as.data.frame(cbind(colData_df, 
                                     spatialCoords_df,
                                     t(intensities_df),
                                     t(nucleus_intensities_df),
                                     t(membrane_intensities_df))
                               ) %&gt;%
  dplyr::rename(cd19 = cd19_opal_480,
                cd68 = cd68_opal_520,
                cd3 = cd3_opal_540, 
                cd8 = cd8_opal_650,
                ier3 = ier3_opal_620,
                pstat3 = p_stat3_opal_570,
                ck = ck_opal_780,
                ki67 = ki67_opal_690) %&gt;%
  # define cell type &#39;immune&#39;
  mutate(immune = ifelse(phenotype_cd19 == &quot;CD19+&quot; | phenotype_cd8 == &quot;CD8+&quot; |
              phenotype_cd3 == &quot;CD3+&quot; | phenotype_cd68 == &quot;CD68+&quot;, &quot;immune&quot;, &quot;other&quot;)) %&gt;%
  dplyr::select(contains(&quot;id&quot;), tissue_category, contains(&quot;phenotype&quot;),
                contains(&quot;position&quot;), ck:dapi, immune) %&gt;%
  # only retain 128 subjects who have clinical data (other 4 are controls)
  dplyr::filter(sample_id %in% patient_level_ovarian$sample_id)


# data frame with clinical characteristics where each row is a different cell
ovarian_df &lt;- full_join(patient_level_ovarian, cell_level_ovarian, by = &quot;sample_id&quot;) %&gt;%
  mutate(sample_id = as.numeric(as.factor(sample_id))) %&gt;%
  filter(tissue_category != &quot;Glass&quot;) %&gt;%
  dplyr::select(sample_id, cell_id, tissue_category, x = cell_x_position, y = cell_y_position,
                everything()) %&gt;%
  select(-tma, -diagnosis, -grade)

rm(oc, assays_slot, intensities_df, nucleus_intensities_df, membrane_intensities_df, colData_df, spatialCoords_df, patient_level_ovarian, cell_level_ovarian)</code></pre>
<p>The resulting dataframe, <code>ovarian_df</code>, contains
information on marker intensities, cell X and Y position, cell
phenotypes, and patient-level characteristics.</p>
</div>
<div id="exploratory-analysis" class="section level3">
<h3>Exploratory analysis</h3>
<pre class="r"><code>dim(ovarian_df)

head(ovarian_df)

glimpse(ovarian_df)</code></pre>
<p>Below we plot the cells for a single subject. In this dataset there
is one image for each subject, and the image comes from a tumor
microarray (TMA).</p>
<pre class="r"><code>id = sample(ovarian_df$sample_id, 1)

ovarian_df %&gt;%
  filter(sample_id == id) %&gt;%
  ggplot(aes(x, y)) +
  geom_point(aes(color = tissue_category), size = 0.1)</code></pre>
<p><img src="MI_VectraPolarisData_files/figure-html/unnamed-chunk-12-1.png" width="90%" /></p>
<p>Now we plot the images (below) for multiple subjects. In this
dataset, multiple TMA cores (each TMA core is frome a different subject)
were placed on the same slide, and then that slide was imaged. X and Y
locations in this data correspond to position on the slide for a given
patient sample.</p>
<pre class="r"><code>ovarian_df %&gt;%
  filter(sample_id &lt; 10) %&gt;%
  ggplot(aes(x, y)) +
  geom_point(aes(color = tissue_category), size = 0.1)</code></pre>
<p><img src="MI_VectraPolarisData_files/figure-html/unnamed-chunk-13-1.png" width="90%" /></p>
</div>
</div>
<div id="lung-data" class="section level2">
<h2>Lung data</h2>
<p>Load the non small cell lung carcinoma data and inspect the
<code>SpatialExperiment</code> object. This dataset has 3-4 images per
patient, representing different ROIs from a tissue slice. In this
dataset, each patient was imaged on a separate slide.</p>
<pre class="r"><code># load lung data
lung &lt;- HumanLungCancerV3()
## see ?VectraPolarisData and browseVignettes(&#39;VectraPolarisData&#39;) for documentation
## loading from cache

# check object
lung
## class: SpatialExperiment 
## dim: 8 1604786 
## metadata(1): clinical_data
## assays(3): intensities nucleus_intensities membrane_intensities
## rownames(8): cd19_opal_650 cd3_opal_520 ... dapi autofluorescence
## rowData names(0):
## colnames: NULL
## colData names(124): cell_id tissue_category ... phenotype_cd4 sample_id
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## spatialCoords names(2) : cell_x_position cell_y_position
## imgData names(0):</code></pre>
<div id="converting-to-dataframe-1" class="section level3">
<h3>Converting to dataframe</h3>
<p>Code below converts the lung cancer dataset from a
<code>SpatialExperiment</code> object to a <code>data.frame</code>.</p>
<pre class="r"><code>## Assays slots
assays_slot &lt;- assays(lung)
intensities_df &lt;- assays_slot$intensities
nucleus_intensities_df&lt;- assays_slot$nucleus_intensities
rownames(nucleus_intensities_df) &lt;- paste0(&quot;nucleus_&quot;, rownames(nucleus_intensities_df))
membrane_intensities_df&lt;- assays_slot$membrane_intensities
rownames(membrane_intensities_df) &lt;- paste0(&quot;membrane_&quot;, rownames(membrane_intensities_df))

# colData and spatialData
colData_df &lt;- colData(lung)
spatialCoords_df &lt;- spatialCoords(lung)

# clinical data
patient_level_lung &lt;- metadata(lung)$clinical_data 

cell_level_lung &lt;- as_tibble(cbind(colData_df, 
                                     spatialCoords_df,
                                     t(intensities_df),
                                     t(nucleus_intensities_df),
                                     t(membrane_intensities_df))
                               )   %&gt;%
  dplyr::rename(cd19 = cd19_opal_650,
                cd3 = cd3_opal_520, 
                cd14 = cd14_opal_540,
                cd8 = cd8_opal_620,
                hladr = hladr_opal_690,
                ck = ck_opal_570) %&gt;%
  dplyr::select(cell_id:slide_id, sample_id:dapi,
                entire_cell_axis_ratio:entire_cell_area_square_microns, contains(&quot;phenotype&quot;))



# data frame with clinical characteristics where each row is a different cell
lung_df &lt;- full_join(patient_level_lung, cell_level_lung, by = &quot;slide_id&quot;) %&gt;%
  #mutate(slide_id = as.numeric(as.factor(slide_id))) %&gt;%
  dplyr::select(image_id = sample_id, patient_id = slide_id, 
                cell_id, x = cell_x_position, y = cell_y_position,
                everything())

rm(lung, assays_slot, intensities_df, nucleus_intensities_df, membrane_intensities_df, colData_df, spatialCoords_df, patient_level_lung,
   cell_level_lung)</code></pre>
<p>The resulting dataframe, <code>lung_df</code>, contains information
on marker intensities, cell X and Y position, cell phenotypes, and
patient-level characteristics.</p>
</div>
<div id="exploratory-analysis-1" class="section level3">
<h3>Exploratory analysis</h3>
<p>Below shows images from a single subject. Each image represents a
different ROI from the same tissue slice from a single subject. There
are 3-6 ROIs for each subject.</p>
<pre class="r"><code>id = sample(lung_df$patient_id, 1)

lung_df %&gt;%
  filter(patient_id == id) %&gt;%
  ggplot(aes(x, y)) +
  geom_point(aes(color = tissue_category), size = 0.1) +
  facet_wrap(~image_id)</code></pre>
<p><img src="MI_VectraPolarisData_files/figure-html/unnamed-chunk-16-1.png" width="90%" /></p>
</div>
</div>
<div id="data-saving" class="section level2">
<h2>Data saving</h2>
<p>Processed data are saved as <code>.rda</code> objects and can also be
downloaded directly from the short course website. For the rest of the
workshop we will be working with this processed data.</p>
<pre class="r"><code># save data sets
save(ovarian_df, file = here::here(&quot;Data&quot;, &quot;ovarian.RDA&quot;))
save(lung_df, file = here::here(&quot;Data&quot;, &quot;lung.RDA&quot;))

# load processed ovarian data
load(url(&quot;https://github.com/julia-wrobel/MI_tutorial/raw/main/Data/ovarian.RDA&quot;))

# load processed lung data
load(url(&quot;https://github.com/julia-wrobel/MI_tutorial/raw/main/Data/lung.RDA&quot;))</code></pre>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<p>Original citations for the two datasets in the
<code>VectraPolarisData</code> package, and the package vignette which
includes a data dicitonary:</p>
<ul>
<li><a
href="https://bioconductor.org/packages/release/data/experiment/vignettes/VectraPolarisData/inst/doc/VectraPolarisData.html">Vignette
for VectraPolarisData</a></li>
<li><a
href="https://aacrjournals.org/mcr/article/19/12/1973/675069/The-Spatial-Context-of-Tumor-Infiltrating-Immune">Paper
for ovarian cancer data</a></li>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/34048945/">Paper for
non-small cell lung carcinoma data</a></li>
</ul>
<p>More resources for <code>SpatialExperiment</code>:</p>
<ul>
<li><a href="https://lmweber.org/OSTA-book/index.html">Orchestrating
Spatially-Resolved Transcriptomics Analysis with Bioconductor</a></li>
<li><a
href="https://pubmed.ncbi.nlm.nih.gov/35482478/">SpatialExperiment:
infrastructure for spatially-resolved transcriptomics data in R using
Bioconductor</a></li>
</ul>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
