<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Tools and software for functional data analysis of multiplexed imaging data</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Statistical analysis of spatial single-cell protein data</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Workshop
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="fridley.html">Overview of multiplex imaging analysis techniques</a>
    </li>
    <li>
      <a href="vandekar.html">Normalization and cell phenotyping for mIF data</a>
    </li>
    <li>
      <a href="chervoneva.html">Quantile index biomarkers for mIF data</a>
    </li>
    <li>
      <a href="wrobel.html">Tools and software for FDA of mIF data</a>
    </li>
    <li>
      <a href="ma.html">A flexible GLM for testing cell-cell colocalization in mIF data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="https://github.com/julia-wrobel/PSB_scProteomics">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Tools and software for functional data
analysis of multiplexed imaging data</h1>

</div>


<div id="slide-deck" class="section level2">
<h2>Slide Deck</h2>
<iframe class="speakerdeck-iframe" frameborder="0" src="https://speakerdeck.com/player/a4680d59269f412faa4a36619abf9644" title="psb_wrobel" allowfullscreen="true" style="border: 0px; background: padding-box padding-box rgba(0, 0, 0, 0.1); margin: 0px; padding: 0px; border-radius: 6px; box-shadow: rgba(0, 0, 0, 0.2) 0px 5px 40px; width: 100%; height: auto; aspect-ratio: 560 / 315;" data-ratio="1.7777777777777777">
</iframe>
</div>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>This code briefly showcases the use of the <code>mxfda</code>
package, an <code>R</code> package for performing functional data
analysis on spatial summary functions from single-cell spatial
proteomics data.</p>
<p>Install <code>mxfda</code> package from GitHub:</p>
<pre class="r"><code>#install devtools if not available
if (!require(&quot;devtools&quot;, quietly = TRUE))
  install.packages(&quot;devtools&quot;)

#install from github
devtools::install_github(&quot;julia-wrobel/mxfda&quot;)</code></pre>
<p>First we load libraries used in this tutorial.</p>
<pre class="r"><code>library(mxfda)
library(tidyverse)
library(patchwork)</code></pre>
<p>Next we load multiplex imaging data. This is a segmented and
phenotyped multiplex immunoflurescence dataset of tumor microarray
images of high grade serous ovarian cancer, used in <a
href="https://aacrjournals.org/mcr/article/19/12/1973/675069/The-Spatial-Context-of-Tumor-Infiltrating-Immune">this
study</a>. The full data are part of the <code>VectraPolarisData</code>
package.</p>
<pre class="r"><code># load processed ovarian cancer data from GitHub
load(url(&quot;https://github.com/julia-wrobel/MI_tutorial/raw/main/Data/ovarian.RDA&quot;))</code></pre>
<pre class="r"><code>ovarian_df &lt;- ovarian_df %&gt;%
  filter(tissue_category == &quot;Tumor&quot;) 

dim(ovarian_df)
## [1] 1292986      35</code></pre>
</div>
<div id="generate-mxfda-object" class="section level2">
<h2>Generate MxFDA object</h2>
<p>The <code>mxfda</code> package uses an S4 object of class
<code>mxFDA</code>. These objects are created with the
<code>make_mxfda()</code> function.</p>
<p>Slots in the <code>mxFDA</code> object are designated as follows:</p>
<ul>
<li><code>Metadata</code> - stores sample specific traits that may be
used as covariates when fitting models</li>
<li><code>Spatial</code> - a data frame of cell level information (x and
y spatial coordinates, phenotype, etc.) that can be used to calculate
spatial summary functions</li>
<li><code>subject_key</code> - a character string for the column in the
metadata that denotes the unique subject id</li>
<li><code>sample_key</code> - a character string for the column in the
metadata that denotes the unique sample id. Note that there may be
multiple samples per subject, and this id links the metadata and spatial
data for each multiplex image sample</li>
<li><code>univariate_summaries</code> and
<code>bivariate_summaries</code> - lists of spatial summary functions
either imported with <code>add_summary_function()</code> or calculated
with <code>extract_summary_functions()</code></li>
<li><code>functional_pca</code> - list of results from functional
principle component analysis</li>
<li><code>functional_mpca</code> - list of results from multilevel
functional principle component analysis</li>
<li><code>functional_cox</code> - list of functional cox models that
have been fit</li>
<li><code>functional_mcox</code> - list of functional mixed cox models
that have been fit</li>
</ul>
<p>Below, we convert the ovarian cancer data to an MxFDA object.</p>
<pre class="r"><code>clinical = ovarian_df %&gt;%
  select(sample_id, age = age_at_diagnosis, survival_time, death, stage_bin) %&gt;%
  distinct()

spatial = ovarian_df %&gt;%
  select(-age_at_diagnosis, -survival_time, -death, -stage_bin)

mxf = make_mxfda(metadata = clinical,
                         spatial = spatial,
                         subject_key = &quot;sample_id&quot;,
                         sample_key = &quot;sample_id&quot;)

rm(ovarian_df, clinical, spatial)</code></pre>
<p>Check out the new MxFDA object:</p>
<pre class="r"><code>summary(mxf)
## mxFDA Object:
##  Subjects: 128
##  Samples: 128
##  Has spatial data
##  Univariate Summaries: None
##  Bivariate Summaries: None
## FPCs not yet calculated
## MFPCs not yet calculated
## FCMs not yet calculated
## MFCMs not yet calculated</code></pre>
<pre class="r"><code>class(mxf)
## [1] &quot;mxFDA&quot;
## attr(,&quot;package&quot;)
## [1] &quot;mxfda&quot;</code></pre>
</div>
<div id="estimate-spatial-summary-functions" class="section level2">
<h2>Estimate spatial summary functions</h2>
<p>You can use the <code>mxfda</code> package to calculate univariate
and bivariate Ripley’s K and nearest neighbor G-functions, or you can
input your own spatial summary functions. Below we calculate univariate
Ripley’s K for immune cells in the dataset.</p>
<pre class="r"><code>mxf = extract_summary_functions(mxf,
                                extract_func = extract_univariate,
                                summary_func = Kest,
                                r_vec = seq(0, 200, by = 1),
                                edge_correction = &quot;iso&quot;,
                                markvar = &quot;immune&quot;,
                                mark1 = &quot;immune&quot;)</code></pre>
<p>Running this code will calculate univariate Ripley’s K function for
immune cells in each image, and will store these spatial summary
functions in the <code>univariate_summaries</code> slot of the S3
<code>mxf</code> object. To access this slot and view the extracted
summary functions, type:</p>
<pre class="r"><code>mxf@univariate_summaries$Kest
## # A tibble: 25,728 × 6
##    sample_id     r sumfun    csr fundiff `immune cells`
##        &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;          &lt;int&gt;
##  1        68     0     0    0       0               198
##  2        68     1     0    3.14   -3.14            198
##  3        68     2     0   12.6   -12.6             198
##  4        68     3   536.  28.3   508.              198
##  5        68     4  1010.  50.3   960.              198
##  6        68     5  1446.  78.5  1368.              198
##  7        68     6  1903. 113.   1790.              198
##  8        68     7  2757. 154.   2603.              198
##  9        68     8  2946. 201.   2745.              198
## 10        68     9  3231. 254.   2976.              198
## # ℹ 25,718 more rows</code></pre>
<p>The summaries are returned as a dataframe. The variable
<code>sumfun</code> is the estimated summary function value,
<code>csr</code> is the theoretical value under complete spatial
randomness, and <code>fundiff</code> =
<code>sumfun</code>-<code>csr</code>; in downstream analysis we will use
the <code>fundiff</code> covariate.</p>
<p>Next, plot the spatial summary functions we just estimated.</p>
<pre class="r"><code>plot(mxf, y = &quot;fundiff&quot;, what = &quot;uni k&quot;) +
  geom_hline(yintercept = 0, color = &quot;red&quot;, linetype = 2)</code></pre>
<p><img src="figsunnamed-chunk-11-1.png" width="90%" /></p>
<p>Values above the dotted red line are interpreted as having more
spatial clustering among immune cells at radius <span
class="math inline">\(r\)</span> than would be expected under complete
spatial randomness.</p>
</div>
<div id="fpca" class="section level2">
<h2>FPCA</h2>
<p>Below we run FPCA on these K-function curves, using the
<code>run_fpca()</code> function, and store these results in the
<code>mxf</code> object.</p>
<pre class="r"><code>mxf &lt;- run_fpca(mxf, 
                metric = &quot;uni k&quot;, 
                r = &quot;r&quot;, 
                value = &quot;fundiff&quot;,
                pve = .99)</code></pre>
<p>Note that the summary of this object now shows the number of
functional principal components (FPCs) that have been calculated:</p>
<pre class="r"><code>mxf
## mxFDA Object:
##  Subjects: 128
##  Samples: 128
##  Has spatial data
##  Univariate Summaries: Kest
##  Bivariate Summaries: None
## FPCs Calculated:
##  Kest: 3 FPCs describe 99.4% variance
## MFPCs not yet calculated
## FCMs not yet calculated
## MFCMs not yet calculated</code></pre>
<p>Below we plot the first two functional principal components from our
FPCA analysis of the K functions.</p>
<pre class="r"><code>p1 = plot(mxf, what = &#39;uni k fpca&#39;, pc_choice = 1)
p2 = plot(mxf, what = &#39;uni k fpca&#39;, pc_choice = 2)

(p1 + p2)</code></pre>
<p><img src="figsfpc_plots-1.png" width="90%" /></p>
<p>The black line in each plot is the mean of the data. In these plots,
the blue and red lines represent one standard deviation of the FPC added
(blue) or subtracted (red) from the population mean.</p>
</div>
<div id="functional-cox-regression" class="section level2">
<h2>Functional Cox regression</h2>
<p>Linear functional Cox regression</p>
<pre class="r"><code>mxf &lt;- run_fcm(mxf, model_name = &quot;fit_lfcm&quot;,
               formula = survival_time ~ age, 
               event = &quot;death&quot;,
               metric = &quot;uni k&quot;, 
               r = &quot;r&quot;, 
               value = &quot;fundiff&quot;,
               afcm = FALSE)</code></pre>
<pre class="r"><code>summary(extract_model(mxf, &#39;uni k&#39;, &#39;fit_lfcm&#39;))
## 
## Family: Cox PH 
## Link function: identity 
## 
## Formula:
## survival_time ~ age + s(t_int, by = l_int * func, bs = &quot;cr&quot;, 
##     k = 20)
## 
## Parametric coefficients:
##     Estimate Std. Error z value Pr(&gt;|z|)
## age  0.01320    0.01123   1.176     0.24
## 
## Approximate significance of smooth terms:
##                       edf Ref.df Chi.sq p-value  
## s(t_int):l_int * func   2  2.001  8.927  0.0115 *
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Deviance explained = 3.35%
## -REML = 357.15  Scale est. = 1         n = 128</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
